openmicroscopy/bioformats
# C++: Add OME-TIFF writer

- Minor refactoring to allow code sharing between reader and writer
- Some additional small warning cleanups
- Add writer and tests

--------

Notable differences from the Java OMETiffWriter implementation:

- `close()` does not have a `canReallyClose` finally block; since by this point we already validated and written out all pixel data and metadata, it's not clear what this is for given that there's nothing we can really do.  What's the rationale for this, given that closing should be terminal and unconditional, particularly resetting the IFD counts which looks like it might screw up the metadata?
- Channel count/Channel objects are fixed up if broken, so some of the special casing isn't needed here, e.g. the equivalent of `planeIndex` in `populateImage`.
- The IFD object isn't currently exposed to the user, so strip/tile sizes etc. aren't settable.  We could either add specific methods to control this, or expose setPlane in the public API (you need to set the current plane before getting the current IFD, since this forces the flushing of the old IFD for which the metadata and pixel data are already set and written).

--------

Testing:

- Build normally and run `make test` or `./bf-test ./cpp/test/ome-bioformats/ometiffwriter` to run just the writer tests.
- Look at cpp/test/ome-bioformats/data/ometiffwriter-*.ome.tiff.  These should all be valid OME-TIFF files which should work with `bf-test info`, `showinf` and `tiffinfo`, and import into OMERO.
- Make the following edit:

```
% git diff
diff --git a/cpp/test/ome-bioformats/ometiffwriter.cpp b/cpp/test/ome-bioformats/ometiffwriter.cpp
index e633d39..e38d3df 100644
--- a/cpp/test/ome-bioformats/ometiffwriter.cpp
+++ b/cpp/test/ome-bioformats/ometiffwriter.cpp
@@ -148,6 +148,7 @@ TEST_P(TIFFWriterTest, setId)
   ome::compat::shared_ptr< ::ome::xml::meta::MetadataRetrieve> retrieve(ome::compat::static_pointer_cast< ::ome::xml::meta::MetadataRetrieve>(meta));
 
   tiffwriter.setMetadataRetrieve(retrieve);
+  tiffwriter.setBigTIFF(true);
 
   bool interleaved = true;
 
```

Run `make && ./cpp/test/ome-bioformats/ometiffwriter` and retest as above.  This should work, but test that it's working with BigTIFF files as well.

Example:

```
% file /tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff 
/tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff: Big TIFF image data, little-endian

% tiffinfo /tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff
TIFF Directory at offset 0x3010 (12304)
  Image Width: 64 Image Length: 64
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Samples/Pixel: 3
  Rows/Strip: 1
  Planar Configuration: single image plane
  ImageDescription: <?xml version="1.0" encoding="UTF-8" standalone="no" ?><OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2013-06" UUID="urn:uuid:7b3b0eb8-686e-48c0-bb00-8c896e64f32e" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2013-06 http://www.openmicroscopy.org/Schemas/OME/2013-06/ome.xsd"><Image ID="Image:0"><Pixels BigEndian="false" DimensionOrder="XYCZT" ID="Pixels:0" Interleaved="true" SignificantBits="8" SizeC="3" SizeT="1" SizeX="64" SizeY="64" SizeZ="1" Type="uint8"><Channel ID="Channel:0:0" SamplesPerPixel="3"/><TiffData FirstC="0" FirstT="0" FirstZ="0" IFD="0" PlaneCount="1"><UUID FileName="ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff">urn:uuid:7b3b0eb8-686e-48c0-bb00-8c896e64f32e</UUID></TiffData></Pixels></Image></OME>
  Software: OME Bio-Formats (C++) 5.1.1

% ./bf-test info /tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff 
Image: /tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff
Using reader: OME-TIFF (Open Microscopy Environment TIFF)
Reader setup took 00:00:00.046091

Filename = "/tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff"
Used files = ["/tmp/b/cpp/test/ome-bioformats/data/ometiffwriter-data-layout-64x64-chunky-strips-01.ome.tiff"]

Reading core metadata
Series count = 1

Series #0:
        Image count = 1
        RGB = [true] ([3]) 
        Interleaved = false
        Indexed = false
        Width = 64
        Height = 64
        SizeZ = 1 (effectively 1)
        SizeT = 1 (effectively 1)
        SizeC = 3 (effectively 1)
        Thumbnail size = 64 × 64
        Endianness = little
        DimensionOrder = XYCZT (certain)
        PixelType = uint8
        Bits per Pixel = 8
        MetadataComplete = true
        ThumbnailSeries = false

No global metadata
```
